pupil_src/shared_modules/surface_tracker/gui.py

```diff
@@ -0,0 +1,568 @@
+            )
+
+            glfw.glfwSetWindowPos(
+                self._window,
+                self.window_position_default[0],
+                self.window_position_default[1],
+            )
+
+            self.trackball = gl_utils.trackball.Trackball()
+            self.input = {"down": False, "mouse": (0, 0)}
+
+            # Register callbacks
+            glfw.glfwSetFramebufferSizeCallback(self._window, self.on_resize)
+            glfw.glfwSetKeyCallback(self._window, self.on_window_key)
+            glfw.glfwSetWindowCloseCallback(self._window, self.on_close)
+            glfw.glfwSetMouseButtonCallback(self._window, self.on_window_mouse_button)
+            glfw.glfwSetCursorPosCallback(self._window, self.on_pos)
+            glfw.glfwSetScrollCallback(self._window, self.on_scroll)
+
+            self.on_resize(self._window, *glfw.glfwGetFramebufferSize(self._window))
+
+            # gl_state settings
+            active_window = glfw.glfwGetCurrentContext()
+            glfw.glfwMakeContextCurrent(self._window)
+            gl_utils.basic_gl_setup()
+            gl_utils.make_coord_system_norm_based()
+
+            # refresh speed settings
+            glfw.glfwSwapInterval(0)
+
+            glfw.glfwMakeContextCurrent(active_window)
+
+    def close_window(self):
+        if self._window:
+            glfw.glfwDestroyWindow(self._window)
+            self._window = None
+            self.window_should_close = False
+
+    def update(self, tex):
+        self.gl_display_in_window(tex)
+
+    def gl_display_in_window(self, world_tex):
+        """
+        here we map a selected surface onto a separate window.
+        """
+        if self._window and self.surface.detected:
+            active_window = glfw.glfwGetCurrentContext()
+            glfw.glfwMakeContextCurrent(self._window)
+            gl_utils.clear_gl_screen()
+
+            # cv uses 3x3 gl uses 4x4 transformation matricies
+            width, height = self.tracker.camera_model.resolution
+            img_corners = np.array(
+                [(0, height), (width, height), (width, 0), (0, 0)], dtype=np.float32
+            )
+            denorm_trans = _get_norm_to_points_trans(img_corners)
+
+            m = self.surface.dist_img_to_surf_trans @ denorm_trans
+            m = gl_utils.cvmat_to_glmat(m)
+
+            gl.glMatrixMode(gl.GL_PROJECTION)
+            gl.glPushMatrix()
+            gl.glLoadIdentity()
+            gl.glOrtho(0, 1, 0, 1, -1, 1)
+            gl.glMatrixMode(gl.GL_MODELVIEW)
+            gl.glPushMatrix()
+            # apply m  to our quad - this will stretch the quad such that the ref surface will span the window extends
+            gl.glLoadMatrixf(m)
+
+            world_tex.draw()
+
+            gl.glMatrixMode(gl.GL_PROJECTION)
+            gl.glPopMatrix()
+            gl.glMatrixMode(gl.GL_MODELVIEW)
+            gl.glPopMatrix()
+
+            # # Draw recent pupil positions onto the surface:
```

Comment:
Imo, this method is quite a bit too long. I think keeping all the ugly gl stuff in this single method is ok, but it would be good to extract `_draw_recent_pupil_positions_onto_surface` or so